core/src/main/scala/cats/data/StateT.scala

```diff
@@ -100,98 +127,131 @@ final class StateT[F[_], S, A](val runF: F[S => F[(S, A)]]) extends Serializable
+    IndexedStateT(s => F.pure((s, s)))
+}
 
-  def lift[F[_], S, A](fa: F[A])(implicit F: Applicative[F]): StateT[F, S, A] =
-    StateT(s => F.map(fa)(a => (s, a)))
+object IndexedStateT extends CommonStateTConstructors with IndexedStateTInstances {
+  def apply[F[_], SA, SB, A](f: SA => F[(SB, A)])(implicit F: Applicative[F]): IndexedStateT[F, SA, SB, A] =
+    new IndexedStateT(F.pure(f))
 
-  def inspect[F[_], S, A](f: S => A)(implicit F: Applicative[F]): StateT[F, S, A] =
-    StateT(s => F.pure((s, f(s))))
+  def applyF[F[_], SA, SB, A](runF: F[SA => F[(SB, A)]]): IndexedStateT[F, SA, SB, A] =
+    new IndexedStateT(runF)
 
-  def inspectF[F[_], S, A](f: S => F[A])(implicit F: Applicative[F]): StateT[F, S, A] =
-    StateT(s => F.map(f(s))(a => (s, a)))
+  def modify[F[_], SA, SB](f: SA => SB)(implicit F: Applicative[F]): IndexedStateT[F, SA, SB, Unit] =
+    IndexedStateT(sa => F.pure((f(sa), ())))
+
+  def modifyF[F[_], SA, SB](f: SA => F[SB])(implicit F: Applicative[F]): IndexedStateT[F, SA, SB, Unit] =
+    IndexedStateT(s => F.map(f(s))(s => (s, ())))
+
+  def set[F[_], SA, SB](sb: SB)(implicit F: Applicative[F]): IndexedStateT[F, SA, SB, Unit] =
+    IndexedStateT(_ => F.pure((sb, ())))
+
+  def setF[F[_], SA, SB](fsb: F[SB])(implicit F: Applicative[F]): IndexedStateT[F, SA, SB, Unit] =
+    IndexedStateT(_ => F.map(fsb)(s => (s, ())))
+}
+
+private[data] abstract class StateTFunctions extends CommonStateTConstructors {
+  def apply[F[_], S, A](f: S => F[(S, A)])(implicit F: Applicative[F]): StateT[F, S, A] =
+    IndexedStateT(f)
+
+  def applyF[F[_], S, A](runF: F[S => F[(S, A)]]): StateT[F, S, A] =
+    IndexedStateT.applyF(runF)
 
   def modify[F[_], S](f: S => S)(implicit F: Applicative[F]): StateT[F, S, Unit] =
-    StateT(s => F.pure((f(s), ())))
+    apply(sa => F.pure((f(sa), ())))
 
   def modifyF[F[_], S](f: S => F[S])(implicit F: Applicative[F]): StateT[F, S, Unit] =
-    StateT(s => F.map(f(s))(s => (s, ())))
-
-  def get[F[_], S](implicit F: Applicative[F]): StateT[F, S, S] =
-    StateT(s => F.pure((s, s)))
+    apply(s => F.map(f(s))(s => (s, ())))
 
   def set[F[_], S](s: S)(implicit F: Applicative[F]): StateT[F, S, Unit] =
-    StateT(_ => F.pure((s, ())))
+    apply(_ => F.pure((s, ())))
 
   def setF[F[_], S](fs: F[S])(implicit F: Applicative[F]): StateT[F, S, Unit] =
-    StateT(_ => F.map(fs)(s => (s, ())))
+    apply(_ => F.map(fs)(s => (s, ())))
 }
 
-private[data] sealed trait StateTInstances extends StateTInstances1 {
-  implicit def catsDataAlternativeForStateT[F[_], S](implicit FM: Monad[F], FA: Alternative[F]): Alternative[StateT[F, S, ?]] =
-    new StateTAlternative[F, S] { implicit def F = FM; implicit def G = FA }
+private[data] sealed trait IndexedStateTInstances extends IndexedStateTInstances1 {
+  implicit def catsDataAlternativeForIndexedStateT[F[_], S](implicit FM: Monad[F], FA: Alternative[F]): Alternative[StateT[F, S, ?]] =
```

Comment:
That's what I meant. Since you are in effect returning an `Alternative with Monad`, why hide the fact that it's a `Monad`? It has a lot more power than the return type suggests.  I just felt a bit odd that your return type isn't the actual public type of the instance. Odd code take a bit more time for code reader to grasp, that's all.